From ec65896c54a26d4947bccd24030442bbbb2d5f6d Mon Sep 17 00:00:00 2001
From: Gabriele Ara <gabriele.ara@santannapisa.it>
Date: Thu, 19 Jan 2023 10:14:00 +0000
Subject: [PATCH 1/3] SCHEDUTIL: consider only DL

---
 kernel/sched/cpufreq_schedutil.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/kernel/sched/cpufreq_schedutil.c b/kernel/sched/cpufreq_schedutil.c
index 831fee509404..17c9f0eb6528 100644
--- a/kernel/sched/cpufreq_schedutil.c
+++ b/kernel/sched/cpufreq_schedutil.c
@@ -210,10 +210,25 @@ unsigned long schedutil_cpu_util(int cpu, unsigned long util_cfs,
 	unsigned long dl_util, util, irq;
 	struct rq *rq = cpu_rq(cpu);
 
+	/*
+	 * Throughout this function I commented some code
+	 * that might pollute our tests on the
+	 * energy-savings of our patches to DEADLINE.
+	 * During our tests, we don't give a damn about
+	 * lower-priority tasks, so long as the kernel does
+	 * not panic we are good.
+	 *
+	 * If you see some code commented out, remember that
+	 * it should NOT be commented out in any real
+	 * environment outside those tests.
+	 */
+
+	/*
 	if (!uclamp_is_used() &&
 	    type == FREQUENCY_UTIL && rt_rq_is_runnable(&rq->rt)) {
 		return max;
 	}
+	*/
 
 	/*
 	 * Early check to see if IRQ/steal time saturates the CPU, can be
@@ -236,9 +251,11 @@ unsigned long schedutil_cpu_util(int cpu, unsigned long util_cfs,
 	 * When there are no CFS RUNNABLE tasks, clamps are released and
 	 * frequency will be gracefully reduced with the utilization decay.
 	 */
+	/*
 	util = util_cfs + cpu_util_rt(rq);
 	if (type == FREQUENCY_UTIL)
 		util = uclamp_util_with(rq, util, p);
+	*/
 
 	dl_util = cpu_util_dl(rq);
 
-- 
2.25.1

