From 47ecf6eb54cf6ea226266aa56a24cbda6c2f1ca3 Mon Sep 17 00:00:00 2001
From: Gabriele Ara <gabriele.ara@santannapisa.it>
Date: Thu, 19 Jan 2023 10:14:34 +0000
Subject: [PATCH 2/3] SCHEDUTIL: use user-defined maximum frequency instead of
 CPUINFO

---
 kernel/sched/cpufreq_schedutil.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/cpufreq_schedutil.c b/kernel/sched/cpufreq_schedutil.c
index 378ae2342208..798e6ad6e0b7 100644
--- a/kernel/sched/cpufreq_schedutil.c
+++ b/kernel/sched/cpufreq_schedutil.c
@@ -171,7 +171,7 @@ static unsigned int get_next_freq(struct sugov_policy *sg_policy,
 {
 	struct cpufreq_policy *policy = sg_policy->policy;
 	unsigned int freq = arch_scale_freq_invariant() ?
-				policy->cpuinfo.max_freq : policy->cur;
+				policy->max : policy->cur;
 
 	freq = map_util_freq(util, freq, max);
 
-- 
2.25.1

