From 1a27fc6de21588f8842e9e3daa0046c3aedbaa9e Mon Sep 17 00:00:00 2001
From: Gabriele Ara <gabriele.ara@santannapisa.it>
Date: Thu, 12 Jan 2023 13:46:50 +0000
Subject: DEADLINE: Fix frequency scaling to use maximum scaling allowed by
 CPUFreq

---
 kernel/sched/deadline.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index d8052c2d87e4..1ff74a6708bd 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -1246,7 +1246,16 @@ static void update_curr_dl(struct rq *rq)
 						 rq,
 						 &curr->dl);
 	} else {
-		unsigned long scale_freq = arch_scale_freq_capacity(cpu);
+		// HOTFIX: taken from arch_topology.c, see function:
+		// topology_set_freq_scale()
+		//
+		// THIS IS NOT THE CORRECT WAY OF CALCULATING THE FREQUENCY
+		// SCALE! THIS IS JUST TO BE USED WHEN THERE IS NOTHING ELSE
+		// AROUND TO GIVE A CORRECT NUMBER FOR THE SCALE!
+		unsigned long cur_freq = cpufreq_quick_get(cpu);
+		unsigned long max_freq = cpufreq_quick_get_max(cpu);
+
+		unsigned long scale_freq = (cur_freq << SCHED_CAPACITY_SHIFT) / max_freq;
 		unsigned long scale_cpu = arch_scale_cpu_capacity(cpu);
 
 		scaled_delta_exec = cap_scale(delta_exec, scale_freq);
-- 
2.25.1

