From a93820518523f6731c016cd5d5662247025a645a Mon Sep 17 00:00:00 2001
From: Gabriele Ara <gabriele.ara@santannapisa.it>
Date: Wed, 25 Jan 2023 13:23:34 +0100
Subject: [PATCH 03/10] Add apedf extraversion

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 094c0f768065..e113012186f8 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 4
 SUBLEVEL = 228
-EXTRAVERSION =
+EXTRAVERSION =-apedf1
 NAME = Kleptomaniac Octopus
 
 # *DOCUMENTATION*
-- 
2.25.1

