From 9cbd275426c91d5e6be98c1498ddf1effb4599e3 Mon Sep 17 00:00:00 2001
From: Gabriele Ara <gabriele.ara@santannapisa.it>
Date: Mon, 31 Jul 2023 09:54:28 +0200
Subject: [PATCH 10/30] [PATCH 03/10] Add apedf extraversion

---
 Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Makefile b/Makefile
index 5dc7e08fa952..2176fd60c08c 100644
--- a/Makefile
+++ b/Makefile
@@ -5,6 +5,8 @@ SUBLEVEL = 249
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
+EXTRAVERSION = -apedf1
+
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
 # More info can be located in ./README
-- 
2.41.0

