From cbda4e5f52c7d4709af872d77b74afc3a52ad6e0 Mon Sep 17 00:00:00 2001
From: luca abeni <luca.abeni@santannapisa.it>
Date: Thu, 18 Nov 2021 13:16:12 +0000
Subject: [PATCH 06/10] Account for the cores' capacity in asymmetric systems

---
 kernel/sched/deadline.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index 9b3ba4e9e51a..6615747378ea 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -2604,6 +2604,7 @@ void init_dl_rq_bw_ratio(struct dl_rq *dl_rq)
 						    global_rt_runtime());
 		dl_rq->max_bw   = to_ratio(global_rt_period(),
 						    global_rt_runtime());
+		dl_rq->max_bw   = dl_rq->max_bw * arch_scale_cpu_capacity(cpu_of(rq_of_dl_rq(dl_rq))) >> SCHED_CAPACITY_SHIFT;
 	}
 }
 
-- 
2.25.1

