From 37e670fb37ee016c2240e8107ce12548d9e1c062 Mon Sep 17 00:00:00 2001
From: luca abeni <luca.abeni@santannapisa.it>
Date: Thu, 18 Nov 2021 13:17:24 +0000
Subject: [PATCH 07/10] Fix bug in push_dl_task: get the task structure when
 needed

---
 kernel/sched/deadline.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index 91b6de51cb27..6fe4fdd854d7 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -2157,18 +2157,16 @@ static int push_dl_task(struct rq *rq)
 	if (!next_task)
 		return 0;
 
-
 	/* We might release rq lock */
+retry:
 	get_task_struct(next_task);
-
-	/* Will lock the rq it'll find */
 	later_rq = find_lock_later_rq_ff(next_task, rq);
 	if (!later_rq) {
-		struct task_struct *task;
-		if (rq->dl.this_bw < rq->dl.max_bw)
+		if (rq->dl.this_bw < rq->dl.max_bw) {
+			put_task_struct(next_task);
 			return 0;
+		}
 
-retry:
 		/* Will lock the rq it'll find */
 		later_rq = find_lock_later_rq(next_task, rq);
 		if (!later_rq) {
-- 
2.40.0

