From 40875ad31de11a1bd329694d650d76d5cdbf285b Mon Sep 17 00:00:00 2001
From: Gabriele Ara <gabriele.ara@santannapisa.it>
Date: Wed, 25 Jan 2023 13:24:22 +0100
Subject: [PATCH 25/30] Change extraversion

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0d95f71726a4..1d5ddac38173 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 4
 SUBLEVEL = 236
-EXTRAVERSION =-apedf1
+EXTRAVERSION =-apedf2
 NAME = Kleptomaniac Octopus
 
 # *DOCUMENTATION*
-- 
2.41.0

