From 241546e4dc23a5f0434b8191881277e55c8c3af1 Mon Sep 17 00:00:00 2001
From: Gabriele Ara <gabriele.ara@santannapisa.it>
Date: Mon, 31 Jul 2023 09:55:19 +0200
Subject: [PATCH 25/30] Change extraversion

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 2176fd60c08c..211fc920ef15 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@ SUBLEVEL = 249
 EXTRAVERSION =
 NAME = Kleptomaniac Octopus
 
-EXTRAVERSION = -apedf1
+EXTRAVERSION = -apedf2
 
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
-- 
2.41.0

