From 6eb62fb349ea6585d8e1abb45d4edb99f6912f4e Mon Sep 17 00:00:00 2001
From: luca abeni <luca.abeni@santannapisa.it>
Date: Thu, 18 Nov 2021 13:10:04 +0000
Subject: [PATCH 02/10] Fix the next bug, in the FF algorithm

To be squashed

See Test4
---
 kernel/sched/deadline.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/deadline.c b/kernel/sched/deadline.c
index 8150dc2c5dc9..28675602869c 100644
--- a/kernel/sched/deadline.c
+++ b/kernel/sched/deadline.c
@@ -2001,8 +2001,9 @@ static int find_later_rq_ff(struct task_struct *task)
 			 "sched RCU must be held");
 	for_each_cpu_and(i, rd->span, cpu_active_mask) {
 		struct rq *rq = cpu_rq(i);
+		u64 added_bw = i == task_cpu(task) ? 0 : task->dl.dl_bw;
 
-		if (cpumask_test_cpu(i, task->cpus_ptr) && (rq->dl.this_bw + task->dl.dl_bw < 1 << BW_SHIFT)) {
+		if (cpumask_test_cpu(i, task->cpus_ptr) && (rq->dl.this_bw + added_bw < 1 << BW_SHIFT)) {
 			return i;
 		}
 	}
-- 
2.25.1

